rfcs/0049-flakes.md

```diff
@@ -0,0 +1,496 @@
+---
+feature: flakes
+start-date: 2019-07-09
+author: Eelco Dolstra
+co-authors: TBD
+shepherd-team: Domen Kožar, Alyssa Ross, Shea Levy, John Ericson
+shepherd-leader: Domen Kožar
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a mechanism to package Nix expressions into
+composable entities called "flakes". Flakes allow hermetic,
+reproducible evaluation of multi-repository Nix projects; impose a
+discoverable, standard structure on Nix projects; and replace previous
+mechanisms such as Nix channels and the Nix search path.
+
+# Motivation
+[motivation]: #motivation
+
+Flakes are motivated by a number of serious shortcomings in Nix:
+
+* While Nix pioneered reproducible builds, sadly, Nix expressions are
+  not nearly as reproducible as Nix builds. Nix expressions can access
+  arbitrary files (such as `~/.config/nixpkgs/config.nix`),
+  environment variables, and Git repositories. This means for instance
+  that it is hard to ensure reproducible evaluation of NixOS or NixOps
+  configurations.
+
+* Nix projects lack discoverability and a standard structure. For
+  example, it's just convention that a repository has a `release.nix`
+  for Hydra jobs and a `default.nix` for packages.
+
+* There is no standard way to compose Nix projects. Typical ways are
+  to rely on the Nix search path (e.g. `import <nixpkgs>`) or to use
+  `fetchGit` or `fetchTarball`. The former has poor reproducibility,
+  while the latter is bad UX because of the need to manually update
+  Git hashes to update dependencies.
+
+* `nix-channel` needs a replacement: channels are hard to create,
+  users cannot easily pin specific versions of channels, channels
+  interact in *ad hoc* ways with the Nix search path, and so on.
+
+The flakes mechanism seeks to address all these problems. This RFC,
+however, only describes the format and semantics of flakes; it doesn't
+describe changes to the `nix` command to support flakes.
+
+# Detailed design
+[design]: #detailed-design
+
+## Flakes
+
+A flake is a directory that contains a file named `flake.nix` in the
+root directory. `flake.nix` specifies some metadata about the flake
+such as dependencies (called *inputs*), as well as its *outputs* (the
+Nix values such as packages or NixOS modules provided by the flake).
+
+As an example, below is the `flake.nix` of
+[`dwarffs`](https://github.com/edolstra/dwarffs) (a FUSE filesystem
+for automatically fetching DWARF debug symbols by ELF build ID). It
+depends on the Nixpkgs flake and provides a package (i.e. an
+installable derivation) and a NixOS module.
+
+```
+{
+  description = "A filesystem that fetches DWARF debug info from the Internet on demand";
+
+  outputs = { self, nixpkgs }: rec {
+    packages.x86_64-linux.dwarffs =
+      with nixpkgs.packages.x86_64-linux;
+      with nixpkgs.builders;
```

Comment:
Just tried this example and it failed with:
```
error: attribute 'builders' missing, at /nix/store/bwa7j11q87qcs02qh9z0wdqc4sc25q4g-source/flake.nix:7:12
```

After that I ran into this issue:
```
error: attribute 'packages' missing, at /nix/store/zi21mr9l9gn20q1an9z455s0fn9zpm54-source/flake.nix:6:12
```

I ended up with the following flake file:
```nix
{
  description = "A filesystem that fetches DWARF debug info from the Internet on demand";

  outputs = { self, nixpkgs }: rec {
    packages.x86_64-linux.dwarffs =
      with nixpkgs.packages.x86_64-linux;
#      with nixpkgs.builders;
      with nixpkgs.lib;

      stdenv.mkDerivation {
        name = "dwarffs-0.1.${substring 0 8 self.lastModifiedDate}";

        buildInputs = [ fuse nix nlohmann_json boost ];

        NIX_CFLAGS_COMPILE = "-I ${nix.dev}/include/nix -include ${nix.dev}/include/nix/config.h -D_FILE_OFFSET_BITS=64";

        src = self;

        installPhase =
          ''
            mkdir -p $out/bin $out/lib/systemd/system

            cp dwarffs $out/bin/
            ln -s dwarffs $out/bin/mount.fuse.dwarffs

            cp ${./run-dwarffs.mount} $out/lib/systemd/system/run-dwarffs.mount
            cp ${./run-dwarffs.automount} $out/lib/systemd/system/run-dwarffs.automount
          '';
      };

#    nixosModules.dwarffs = ...;

    defaultPackage.x86_64-linux = packages.x86_64-linux.dwarffs;

    checks.build = packages.x86_64-linux.dwarffs;
  };
}
```

Steps to reproduce:

1. Spawn a nix-shell with `nixFlakes`: `nix-shell -p nixFlakes` (on recent/current nixos-unstable channel, commit 029a5de0839 in my case)
2. Enable the `flakes` and `nix-command` feature in ~/.config/nix/nix.conf (verify via `nix show-config | grep experi`)
3. Write the above `flake.nix` into a new (empty?) directory
4. change to that new directory and execute `nix build flake:.`

Maybe it is worth exchanging the example with something simpler. How about a variant of the example(s) presented in the blog post?